File: 2023/04. Scratchcards/c++/two/src/main.cpp
#include <algorithm>
#include <cstddef>
#include <fstream>
#include <iomanip>
#include <iostream>
#include <limits>
#include <string>
#include <vector>

int main(const int argc, const char** const argv)
{
    if (argc != 2)
    {
        std::cerr << "Usage: " << argv[0] << " <INPUT_FILE>\n";
        return 1;
    }

    std::ifstream infile{argv[1]};
    if (!infile.is_open())
    {
        std::cerr << "ERROR! Failed to open " << std::quoted(argv[1]) << '\n';
        return 1;
    }

    // Puzzle result
    int cardCount{};

    // Numbers found and winning numbers respectively. Both cleared after each iteration
    std::vector<int> presentNumbers;
    std::vector<int> winningNumbers;

    // Number of extra copies to add for the next card is present in front
    std::vector<int> copies;

    std::size_t lineNum{1};
    for (std::string line; std::getline(infile, line); ++lineNum, presentNumbers.clear(), winningNumbers.clear())
    {
        std::istringstream ss{line};
        
        // Check for "Card" string
        ss >> line;
        if (line != "Card")
        {
            std::cerr << "Line " << lineNum << " is invalied\n";
            return 1;
        }

        // Check card number matches expected
        int cardNum;
        ss >> cardNum;
        if (cardNum != static_cast<int>(lineNum))
        {
            std::cerr << "Invalid card number on line " << lineNum << '\n';
            return 1;
        }

        // Check for colon to mark start of present numbers
        if (ss.get() != ':')
        {
            std::cerr << "Expected colon on line " << lineNum << '\n';
            return 1;
        }

        // Read all present numbers. '|' triggers failure that will be cleared after loop
        for (int number; ss >> number;)
        {
            presentNumbers.push_back(number);
        }
        ss.clear();

        // Find and remove pipe
        ss.ignore(std::numeric_limits<std::streamsize>::max(), '|');

        // Read all winning numbers
        for (int number; ss >> number;)
        {
            winningNumbers.push_back(number);
        }

        // Sort for binary search for fast lookup
        std::ranges::sort(winningNumbers);

        // Count wins of card
        std::size_t wins{};
        for (const int presentNumber : presentNumbers)
        {
            if (std::ranges::binary_search(winningNumbers, presentNumber))
                ++wins;
        };

        // Add extra copies of necessary
        int copiesCurrent{1};
        if (!copies.empty())
        {
            copiesCurrent += copies.front();
            copies.erase(copies.begin());
        }

        // Add copies to total card count
        cardCount += copiesCurrent;

        // Resize copies buffer to ensure enough room them increment copy counts
        copies.resize(std::max(wins, copies.size()));
        for (std::size_t i{}; i < wins; ++i)
        {
            copies[i] += copiesCurrent;
        }
    }

    std::cout << "Result: " << cardCount << '\n';

    return 0;
}

File: 2023/04. Scratchcards/rust/two/src/main.rs
use std::process::ExitCode;
use std::{env, fs};

fn main() -> ExitCode {
    let args: Vec<String> = env::args().collect();
    if args.len() != 2 {
        eprintln!("Usage: {} <INPUT_FILE>", args[0]);
        return ExitCode::from(1);
    }

    // Puzzle result
    let mut card_count = 0;

    // Numbers found and winning numbers respectively. Both cleared after each iteration
    let mut present_numbers = Vec::new();
    let mut winning_numbers = Vec::new();

    // Number of extra copies to add for the next card is present in front
    let mut copies = Vec::new();

    // Read each line
    let contents = fs::read_to_string(&args[1]).expect("Failed to load input file");
    for (line_num, line) in contents.lines().enumerate() {
        present_numbers.clear();
        winning_numbers.clear();

        let mut tokens = line.split(' ').filter(|token| !token.is_empty());

        // Read and verify "Card"
        if tokens.next() != Some("Card") {
            eprintln!("ERROR! Expected \"Card\" on line {line_num}");
            return ExitCode::from(1);
        }

        // Read and verify card number
        if tokens.next() != Some(&format!("{}:", line_num + 1)) {
            eprintln!("ERROR! Wrong card number on line {line_num}");
            return ExitCode::from(1);
        }

        // Reference to Vec to store numbers in
        let mut target = &mut present_numbers;

        for token in tokens {
            if token == "|" {
                // Switch from reading present numbers to winning numbers
                target = &mut winning_numbers;
            } else {
                // Parse number
                let number: i32 = token.parse().expect(&format!(
                    "Invlaid number on line {}: \"{token}\"",
                    line_num + 1
                ));

                target.push(number);
            }
        }

        // Sort for binary search
        winning_numbers.sort();

        let mut wins = 0;

        // Calculate card score
        for present_number in present_numbers.iter() {
            if let Ok(_) = winning_numbers.binary_search(present_number) {
                wins += 1;
            }
        }

        // Add extra copies of necessary
        let copies_current = 1
            + (!copies.is_empty())
                .then(|| copies.remove(0))
                .unwrap_or_default();

        // Add copies to total card count
        card_count += copies_current;

        // Resize copies buffer to ensure enough room them increment copy counts
        copies.resize(wins.max(copies.len()), 0);
        for copy_count in &mut copies[..wins] {
            *copy_count += copies_current;
        }
    }

    println!("Result: {card_count}");

    ExitCode::from(0)
}
